job/deployer.go

```diff
@@ -120,14 +146,117 @@ func (d *deployer) cleanPerNamespace(ctx context.Context, namespaceSpec models.N
 	return nil
 }
 
-func (d *deployer) enrichJobSpecWithHookDependencies(jobSpecs []models.JobSpec) []models.JobSpec {
-	var enrichedJobSpecs []models.JobSpec
-	for _, jobSpec := range jobSpecs {
-		hooks := d.dependencyResolver.FetchHookWithDependencies(jobSpec)
-		if len(hooks) > 0 {
-			jobSpec.Hooks = hooks
+func (d *deployer) enrichJobSpecs(ctx context.Context, jobSpecs []models.JobSpec, deploymentProjectSpec models.ProjectSpec) error {
+	jobIDInferredDependenciesMap, err := d.getJobIDInferredDependenciesMap(ctx, deploymentProjectSpec, jobSpecs)
+	if err != nil {
+		return err
+	}
+
+	for i, job := range jobSpecs {
+		targetJobSpec := job
+		if err := d.enrichWithStaticDependencies(ctx, &targetJobSpec, deploymentProjectSpec); err != nil {
+			return fmt.Errorf("error while enriching jobspec %d with static dependencies: %w", targetJobSpec.ID, err)
+		}
+
+		inferredDependencies := jobIDInferredDependenciesMap[job.ID]
+		d.enrichWithInferredDependencies(&targetJobSpec, deploymentProjectSpec, inferredDependencies)
+
+		d.enrichWithHookDependencies(&targetJobSpec)
+
+		jobSpecs[i] = targetJobSpec
+	}
+	return nil
+}
+
+func (d *deployer) enrichWithStaticDependencies(
+	ctx context.Context,
+	jobSpec *models.JobSpec,
+	deploymentProjectSpec models.ProjectSpec,
+) error {
+	projectJobSpecRepo := d.projectJobSpecRepoFactory.New(deploymentProjectSpec)
+	staticDependencies, err := d.dependencyResolver.ResolveStaticDependencies(ctx, *jobSpec, deploymentProjectSpec, projectJobSpecRepo)
+	jobSpec.Dependencies = staticDependencies
+	return err
+}
+
+func (*deployer) enrichWithInferredDependencies(
+	jobSpec *models.JobSpec,
+	deploymentProjectSpec models.ProjectSpec,
+	resourceDependencies []models.JobSpec,
+) {
+	if jobSpec.Dependencies == nil {
+		jobSpec.Dependencies = make(map[string]models.JobSpecDependency)
+	}
+	for _, dependency := range resourceDependencies {
+		dependencyType := models.JobSpecDependencyTypeIntra
+		if dependency.NamespaceSpec.ProjectSpec.ID.UUID() != deploymentProjectSpec.ID.UUID() {
+			dependencyType = models.JobSpecDependencyTypeInter
+		}
+		jobDependencySpec := dependency
+		jobSpec.Dependencies[dependency.Name] = models.JobSpecDependency{
+			Project: &jobDependencySpec.NamespaceSpec.ProjectSpec,
+			Job:     &jobDependencySpec,
+			Type:    dependencyType,
+		}
+	}
+}
+
+func (d *deployer) enrichWithHookDependencies(jobSpec *models.JobSpec) {
+	hooks := d.dependencyResolver.FetchHookWithDependencies(*jobSpec)
+	jobSpec.Hooks = hooks
+}
+
+func (d *deployer) getJobIDInferredDependenciesMap(ctx context.Context, deploymentProjectSpec models.ProjectSpec,
```

Comment:
what this method is doing is ambigious, lets discuss again.

Reply:
on the private methods too

Reply:
this is simplified by moving the complexity to the repository and only by selecting needed fields, not the whole job spec for each dependency.